src/config/id.rs

```diff
@@ -0,0 +1,106 @@
+use serde::de::{self, Visitor};
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::cmp::{Ord, Ordering, PartialOrd};
+use std::fmt;
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct ComponentId {
+    pub id: String,
+}
+
+impl ComponentId {
+    pub fn global<T: Into<String>>(id: T) -> Self {
+        Self { id: id.into() }
+    }
+}
+
+impl From<String> for ComponentId {
+    fn from(id: String) -> Self {
+        Self { id }
+    }
+}
+
+impl From<&str> for ComponentId {
+    fn from(id: &str) -> Self {
+        Self { id: id.to_string() }
+    }
+}
```

Comment:
According to rustc, `impl<T: ToString> From<T> for ComponentId` is conflicting with `impl From<T> to T`, so apparently it's not possible because `ComponentId` implements `ToString`.